src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -4097,12 +4105,29 @@ void AdaptDispatch::_ReportSGRSetting() const
     addAttribute(L";1"sv, attr.IsIntense());
     addAttribute(L";2"sv, attr.IsFaint());
     addAttribute(L";3"sv, attr.IsItalic());
-    addAttribute(L";4"sv, attr.IsUnderlined());
+
+    const auto ulStyle = attr.GetUnderlineStyle();
+    switch (ulStyle)
+    {
+    case UnderlineStyle::SinglyUnderlined:
+        addAttribute(L";4"sv, true);
+        break;
+    case UnderlineStyle::DoublyUnderlined:
+        addAttribute(L";21"sv, true);
+        break;
+    case UnderlineStyle::CurlyUnderlined:
+    case UnderlineStyle::DottedUnderlined:
+    case UnderlineStyle::DashedUnderlined:
+        addAttribute(fmt::format(FMT_COMPILE(";4:{}"sv), WI_EnumValue(ulStyle)), true);
+        break;
+    default:
+        break;
+    }
```

Comment:
Could we not have an `IsUnderlined` method that takes an `UnderlineStyle`, and then this could more easily follow the same pattern as all the other attributes? Something like this:
```
addAttribute(L";4"sv, attr.IsUnderlined(UnderlineStyle::SinglyUnderlined));
addAttribute(L";4:3"sv, attr.IsUnderlined(UnderlineStyle::CurlyUnderlined));
addAttribute(L";4:4"sv, attr.IsUnderlined(UnderlineStyle::DottedUnderlined));
addAttribute(L";4:5"sv, attr.IsUnderlined(UnderlineStyle::DashedUnderlined));
addAttribute(L";21"sv, attr.IsUnderlined(UnderlineStyle::DoublyUnderlined));
```
This way we could more easily keep the rendition attributes in their numeric order, so the doubly underline test would remain between `IsCrossedOut` and `IsOverlined`.

Reply:
I tried to avoid having `IsUnderlined(UnderlineStyle)` 😅 because it gives false idea that testing each style is 'trivial', but actually, it calls `_CharacterAttrsToUnderlineStyle(...)` everytime.

It would be better if we do that through `underlineStyle.IsUnderlined(...)`, WDYT?

Reply:
I tried to avoid having IsUnderlined(UnderlineStyle) sweat_smile because it gives false idea that testing each style is 'trivial', but actually, it calls _CharacterAttrsToUnderlineStyle(...) everytime.

It would be better if we do that through underlineStyle.IsUnderlined(...), WDYT?

Reply:
In most cases switch/case would still be a better choice. But maybe here we might do it through a lambda which would abstract that out?

Reply:
Instead of using `_CharacterAttrToUnderlineStyle`, you could implement `IsUnderlined` as something like this:

```
(_attr & CharacterAttributes::UnderlineStyle) == (style << UNDERLINE_STYLE_SHIFT);
```

The shift would then be evaluated at compile time and it should end up being a simple test of the form `(_attr & 0x1c0) == 0x20`. And if the optimizer is any good, the `_attr & 0x1c0` would only be executed once in a sequence of tests like we have here.

That said, I think it's fine as you have it in the latest commits.

Reply:
I'm planning to do something about this function, probably will do what you suggested.